In [1]:
from pathlib import Path
import pandas as pd

project_dir = Path(r"C:\Users\HP-ZBOOK i7\graphrag_test\pilot_01")
csv_path = project_dir / "inspection" / "01_entities_all.csv"

entities = pd.read_csv(csv_path)

print("Number of entities:", len(entities))
print("Columns:", entities.columns.tolist())
print()
print(entities["type"].value_counts())

Number of entities: 312
Columns: ['id', 'human_readable_id', 'title', 'type', 'description', 'text_unit_ids', 'frequency', 'degree']

type
ORGANIZATION    142
EVENT           121
PERSON           25
GEO              24
Name: count, dtype: int64


In [2]:
pd.set_option("display.max_colwidth", 120)

entities[
    ["human_readable_id", "title", "type", "frequency", "degree"]
].head(30)

,human_readable_id,title,type,frequency,degree
0,0,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,EVENT,14,39
1,1,"BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENERGIE, MOBILITÄT, INNOVATION UND TECHNOLOGIE",ORGANIZATION,2,9
2,2,HUBERT FECHNER,PERSON,1,2
3,3,LEONORE GEWESSLER,PERSON,1,2
4,4,ÖSTERREICH,GEO,15,82
5,5,WIEN,GEO,3,8
6,6,ERNEUERBAREN-AUSBAU-GESETZ (EAG),EVENT,2,5
7,7,INTEGRIERTER ÖSTERREICHISCHER NETZINFRASTRUKTURPLAN (ÖNIP),EVENT,1,1
8,8,PHOTOVOLTAIK (PV),EVENT,1,4
9,9,CAJETAN PERWEIN,PERSON,1,1


In [3]:
entities[
    ["title", "type", "frequency", "degree", "description"]
].sort_values(
    by=["degree", "frequency"],
    ascending=False
).head(25)

,title,type,frequency,degree,description
4,ÖSTERREICH,GEO,15,82,"Österreich, also known as Austria, is a Central European nation and a member of the European Union that is at the fo..."
0,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,EVENT,14,39,"The ""Österreichische Photovoltaik-Strategie"" (Austrian Photovoltaic Strategy) is Austria’s official national strateg..."
38,PV-ANLAGE,ORGANIZATION,3,20,PV-ANLAGE (Photovoltaic system) refers to decentralized solar power generation installations that are widely deploye...
10,BMK,ORGANIZATION,7,16,"BMK, the commonly used abbreviation for the Austrian Federal Ministry for Climate Action, Environment, Energy, Mobil..."
13,KLIMA- UND ENERGIEFONDS,ORGANIZATION,6,14,The **KLIMA- UND ENERGIEFONDS** (Climate and Energy Fund) is an Austrian governmental and public organization dedica...
43,ENERGIEGEMEINSCHAFTEN,ORGANIZATION,6,14,ENERGIEGEMEINSCHAFTEN (energy communities) are formal or informal organizations or associations composed of citizens...
78,BUND,ORGANIZATION,1,14,"The federal government of Austria, which collaborates with the states (Bundesländer) to achieve national renewable e..."
56,PHOTOVOLTAIKANLAGEN,ORGANIZATION,2,13,"Photovoltaikanlagen (Photovoltaic Installations) refers collectively to all organizations, firms, and actors involve..."
21,E-CONTROL,ORGANIZATION,6,11,E-CONTROL is the official Austrian regulatory authority tasked with overseeing and managing the country's electricit...
37,BUNDESLÄNDER,GEO,4,11,"The ""Bundesländer"" are the federal states within Austria, functioning as subnational administrative regions. Each Bu..."


In [4]:
# Twenty most-connected entities
top_connected = entities.nlargest(20, "degree")

# Ten reproducibly sampled entities from each raw GraphRAG type
random_by_type = (
    entities.groupby("type", group_keys=False)
    .sample(n=10, random_state=42)
)

# Combine the groups and remove any repeated rows
entity_audit_sample = (
    pd.concat([top_connected, random_by_type])
    .drop_duplicates(subset="id")
    .reset_index(drop=True)
)

print("Audit sample size:", len(entity_audit_sample))
print()
print(entity_audit_sample["type"].value_counts())

entity_audit_sample[
    ["title", "type", "frequency", "degree", "description", "text_unit_ids"]
]

Audit sample size: 57

type
ORGANIZATION    22
EVENT           13
GEO             12
PERSON          10
Name: count, dtype: int64


,title,type,frequency,degree,description,text_unit_ids
0,ÖSTERREICH,GEO,15,82,"Österreich, also known as Austria, is a Central European nation and a member of the European Union that is at the fo...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...
1,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,EVENT,14,39,"The ""Österreichische Photovoltaik-Strategie"" (Austrian Photovoltaic Strategy) is Austria’s official national strateg...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...
2,PV-ANLAGE,ORGANIZATION,3,20,PV-ANLAGE (Photovoltaic system) refers to decentralized solar power generation installations that are widely deploye...,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...
3,BMK,ORGANIZATION,7,16,"BMK, the commonly used abbreviation for the Austrian Federal Ministry for Climate Action, Environment, Energy, Mobil...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...
4,KLIMA- UND ENERGIEFONDS,ORGANIZATION,6,14,The **KLIMA- UND ENERGIEFONDS** (Climate and Energy Fund) is an Austrian governmental and public organization dedica...,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...
5,ENERGIEGEMEINSCHAFTEN,ORGANIZATION,6,14,ENERGIEGEMEINSCHAFTEN (energy communities) are formal or informal organizations or associations composed of citizens...,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...
6,BUND,ORGANIZATION,1,14,"The federal government of Austria, which collaborates with the states (Bundesländer) to achieve national renewable e...",['23663988970639bea940978ccbd9150fd292856f2f31a1d587392606a64713d5dfeade66f1beea67d1cc0b0e93012337685e0eca3267f98c16...
7,PHOTOVOLTAIKANLAGEN,ORGANIZATION,2,13,"Photovoltaikanlagen (Photovoltaic Installations) refers collectively to all organizations, firms, and actors involve...",['f9e54ba797ff2b80a5f0e5f24e3901236ac4130e4e84d01615e67bc810b290cd5ac7c08f71a782797073050d512eda2695c3659569deba0d4b...
8,E-CONTROL,ORGANIZATION,6,11,E-CONTROL is the official Austrian regulatory authority tasked with overseeing and managing the country's electricit...,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...
9,BUNDESLÄNDER,GEO,4,11,"The ""Bundesländer"" are the federal states within Austria, functioning as subnational administrative regions. Each Bu...",['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...


In [5]:
entity_review = entity_audit_sample[
    [
        "id",
        "title",
        "type",
        "description",
        "text_unit_ids",
        "frequency",
        "degree",
    ]
].copy()

entity_review["source_supported"] = ""
entity_review["primary_classification"] = ""
entity_review["expected_type"] = ""
entity_review["duplicate_or_alias_of"] = ""
entity_review["description_problem"] = ""
entity_review["likely_pipeline_stage"] = ""
entity_review["audit_notes"] = ""

entity_review.head()

,id,title,type,description,text_unit_ids,frequency,degree,source_supported,primary_classification,expected_type,duplicate_or_alias_of,description_problem,likely_pipeline_stage,audit_notes
0,05c9af58-ac32-4ec6-b923-358b587e8cae,ÖSTERREICH,GEO,"Österreich, also known as Austria, is a Central European nation and a member of the European Union that is at the fo...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...,15,82,,,,,,,
1,cb0a1b84-f9a3-4404-9807-afa3d1e6c3af,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,EVENT,"The ""Österreichische Photovoltaik-Strategie"" (Austrian Photovoltaic Strategy) is Austria’s official national strateg...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...,14,39,,,,,,,
2,87671c51-7ba4-4fb4-97a0-6e4dde071aaf,PV-ANLAGE,ORGANIZATION,PV-ANLAGE (Photovoltaic system) refers to decentralized solar power generation installations that are widely deploye...,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...,3,20,,,,,,,
3,a904e0ff-8af4-4090-9452-b6ee49f26cb8,BMK,ORGANIZATION,"BMK, the commonly used abbreviation for the Austrian Federal Ministry for Climate Action, Environment, Energy, Mobil...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...,7,16,,,,,,,
4,3f8fbdde-b872-4a1e-b5d9-dd3c12a82852,KLIMA- UND ENERGIEFONDS,ORGANIZATION,The **KLIMA- UND ENERGIEFONDS** (Climate and Energy Fund) is an Austrian governmental and public organization dedica...,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...,6,14,,,,,,,


## Stage 3 — Preliminary Entity-Extraction Inspection

### Objective

The purpose of this stage is to evaluate whether GraphRAG extracted appropriate entities from the Austrian Photovoltaic Strategy and whether those entities were assigned suitable types.

The entity audit distinguishes between:

1. **Entity-identification quality** — whether GraphRAG identified a meaningful concept from the source.
2. **Entity-typing quality** — whether GraphRAG assigned an appropriate category to that concept.
3. **Entity-resolution quality** — whether different names for the same concept were merged.
4. **Description quality** — whether the LLM-generated entity description is supported by the source.

This distinction is important because an entity may exist in the source but still have the wrong type, an unsupported description, or an unresolved duplicate.

---

### Complete entity output

GraphRAG extracted **312 entities** from the single Austrian Photovoltaic Strategy document.

The raw type distribution was:

| Raw GraphRAG type |   Count |    Share |
| ----------------- | ------: | -------: |
| `ORGANIZATION`    |     142 |    45.5% |
| `EVENT`           |     121 |    38.8% |
| `PERSON`          |      25 |     8.0% |
| `GEO`             |      24 |     7.7% |
| **Total**         | **312** | **100%** |

The graph-extraction configuration allowed only four entity types:

```yaml
entity_types:
  - organization
  - person
  - geo
  - event
```

These categories do not adequately represent the concepts found in the Austrian solar-market document. Missing domain-relevant categories include:

* `POLICY`
* `LAW`
* `TECHNOLOGY`
* `INFRASTRUCTURE`
* `TARGET`
* `SUPPORT_SCHEME`
* `CONSTRAINT`
* `MARKET_METRIC`
* `STAKEHOLDER_GROUP`
* `PROCESS`

The large number of entities classified as `ORGANIZATION` or `EVENT` must therefore not be interpreted as evidence that the document primarily describes organizations and events. The distribution is strongly influenced by the limited configuration.

---

### Sampling method

Manually reviewing all 312 entities in detail would be time-consuming. A systematic sample was therefore constructed.

The sample combines:

* The **20 entities with the highest graph degree**
* A reproducible random sample of **10 entities from each raw GraphRAG type**
* Removal of duplicated rows appearing in both groups

A fixed random seed was used:

```python
random_state=42
```

This makes the sample reproducible.

After removing overlaps, the audit sample contains **57 unique entities**:

| Raw type       | Sample count |
| -------------- | -----------: |
| `ORGANIZATION` |           22 |
| `EVENT`        |           13 |
| `GEO`          |           12 |
| `PERSON`       |           10 |
| **Total**      |       **57** |

The sample includes both highly connected entities and less prominent entities from every available raw type. It therefore avoids reviewing only the most obvious or important examples.

---

### Meaning of GraphRAG fields

The entity table includes:

| Field           | Meaning                                                  |
| --------------- | -------------------------------------------------------- |
| `title`         | Entity name produced by GraphRAG                         |
| `type`          | Raw entity type assigned by GraphRAG                     |
| `description`   | LLM-generated summary of the entity                      |
| `text_unit_ids` | Text units associated with the entity                    |
| `frequency`     | Number of extraction occurrences combined for the entity |
| `degree`        | Number of graph relationships connected to the entity    |

`frequency` and `degree` are structural measurements. They are not measures of factual accuracy or confidence.

For example, `ÖSTERREICH` has a degree of 82. This means that GraphRAG connected Austria to many other entities. It does not establish that all 82 relationships are correct.

---

## Preliminary observations

The first inspection already reveals substantial entity-typing and entity-resolution problems. These findings are preliminary because the complete 57-entity sample has not yet been individually classified against the source.

### Apparently correct examples

Some entities were identified and typed sensibly:

| Entity                     | Raw type       | Preliminary assessment    |
| -------------------------- | -------------- | ------------------------- |
| `ÖSTERREICH`               | `GEO`          | Correct geographic entity |
| `WIEN`                     | `GEO`          | Correct geographic entity |
| `E-CONTROL`                | `ORGANIZATION` | Correct organization      |
| `EUROPÄISCHE KOMMISSION`   | `ORGANIZATION` | Correct organization      |
| `KLIMA- UND ENERGIEFONDS`  | `ORGANIZATION` | Correct organization      |
| `BUNDESVERBAND PV AUSTRIA` | `ORGANIZATION` | Correct organization      |
| `LEONORE GEWESSLER`        | `PERSON`       | Correct person            |
| `HACKNER, JOHANNES`        | `PERSON`       | Correct person            |

These examples indicate that GraphRAG can successfully identify conventional named entities that fit its available type system.

---

### Wrong entity types

Many source-supported concepts do not fit the four available GraphRAG categories and were assigned unsuitable types.

Examples include:

| Entity                                   | GraphRAG type  | More suitable type                   |
| ---------------------------------------- | -------------- | ------------------------------------ |
| `ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE` | `EVENT`        | `POLICY` or `STRATEGY`               |
| `PV-ANLAGE`                              | `ORGANIZATION` | `TECHNOLOGY` or `INFRASTRUCTURE`     |
| `PHOTOVOLTAIKANLAGEN`                    | `ORGANIZATION` | `TECHNOLOGY` or `INFRASTRUCTURE`     |
| `STROMNETZ`                              | `ORGANIZATION` | `INFRASTRUCTURE`                     |
| `ENERGIEWENDE`                           | `EVENT`        | `PROCESS` or strategic concept       |
| `STROMMARKTPREISE`                       | `EVENT`        | `MARKET_METRIC`                      |
| `BIODIVERSITÄT`                          | `EVENT`        | Environmental factor or `CONSTRAINT` |
| `REGIERUNGSPROGRAMM`                     | `EVENT`        | `POLICY`                             |
| `AKTIONSPLAN NETZANSCHLUSS 2023`         | `EVENT`        | `POLICY` or action plan              |
| `RICHTLINIE 98/70/EG`                    | `EVENT`        | `LAW` or regulation                  |
| `WASSERSTOFF`                            | `ORGANIZATION` | `TECHNOLOGY` or energy carrier       |
| `BIOENERGIE`                             | `ORGANIZATION` | `TECHNOLOGY` or energy source        |
| `VERKEHRSSEKTOR`                         | `ORGANIZATION` | `SECTOR`                             |
| `BEVÖLKERUNG`                            | `PERSON`       | `STAKEHOLDER_GROUP`                  |
| `DACHDECKER:INNEN`                       | `PERSON`       | `STAKEHOLDER_GROUP` or occupation    |
| `PRIVATE INVESTOREN`                     | `PERSON`       | `STAKEHOLDER_GROUP`                  |
| `PRIVATPERSONEN`                         | `PERSON`       | `STAKEHOLDER_GROUP`                  |
| `FREIFLÄCHEN`                            | `GEO`          | `LAND_USE` or site type              |
| `REIHENHÄUSER`                           | `GEO`          | `BUILDING_TYPE`                      |
| `BETRIEBSANLAGEN`                        | `GEO`          | `INFRASTRUCTURE` or site type        |

These examples suggest that GraphRAG frequently identified a real concept but assigned the wrong type because the configuration did not provide an appropriate category.

This is primarily an **entity-type/schema failure**, not necessarily an entity-identification failure.

---

### Duplicate and alias problems

The preliminary inspection also reveals several likely cases where the same real-world concept was represented by multiple nodes.

#### Austrian Photovoltaic Strategy

Possible aliases include:

```text
ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE
PHOTOVOLTAIK-STRATEGIE
PV-STRATEGIE
```

These appear to refer to the same strategy and should probably map to one canonical entity.

#### Austrian federal ministry

Possible aliases include:

```text
BMK
BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENERGIE,
MOBILITÄT, INNOVATION UND TECHNOLOGIE
```

The abbreviation and full name should be connected through an alias relationship or normalized to one canonical organization.

#### Photovoltaic installations

Possible variants include:

```text
PV-ANLAGE
PV-ANLAGEN
PHOTOVOLTAIKANLAGEN
```

These may represent singular, plural, or language variants of the same basic concept. Their precise treatment may depend on whether the final KG distinguishes between the technology class and individual installations.

#### Integrated Austrian Network Infrastructure Plan

Possible variants include:

```text
INTEGRIERTER ÖSTERREICHISCHER NETZINFRASTRUKTURPLAN (ÖNIP)
INTEGRIERTER ÖSTERREICHISCHER NETZINFRASTRUKTURPLAN (NIP)
```

These require verification because `ÖNIP` is the abbreviation used for the Austrian plan, while `NIP` may be an inconsistent variant.

#### Johannes Hackner

Possible variants include:

```text
HACKNER, JOHANNES
JOHANNES HACKNER
```

These appear to be different name orders for the same person.

These cases indicate a **record-linkage/entity-resolution problem**. Adding more entity types alone will not automatically merge abbreviations, singular/plural forms, shortened names, or reversed personal names.

---

### Overgeneralized entities

Some extracted nodes describe broad categories rather than clearly bounded entities.

Examples include:

```text
BUND
BEVÖLKERUNG
NATUR
MITGLIEDSTAATEN
VERORDNUNGEN
ENERGIEWENDE
PV
```

Some of these concepts may still be useful in a controlled KG, but they require explicit modelling decisions.

For example:

* `PV` may be an alias for the technology `Photovoltaics`.
* `ENERGIEWENDE` may be a strategic process rather than an event.
* `BEVÖLKERUNG` may be represented as a stakeholder category.
* `VERORDNUNGEN` may be too general unless linked to specific regulations.

The audit should not automatically reject every broad concept. It should determine whether the concept supports the project’s purpose of explaining relationships among Austrian solar-market policy, market, and technical factors.

---

### Description-quality concerns

GraphRAG’s entity descriptions were generated by an LLM and may combine information from several chunks.

Some descriptions use broad or evaluative language such as:

* “at the forefront”
* “plays a crucial role”
* “comprehensive national strategy”
* “widely deployed”
* “central hub”
* “leading role”

Such language may be reasonable as a summary but must not automatically be treated as an explicit source statement.

An entity can therefore have:

* A correct name
* A correct or incorrect type
* A partially unsupported description

Description quality must be assessed separately from entity existence.

---

## Initial diagnosis

The preliminary evidence suggests:

1. GraphRAG can identify conventional named entities such as countries, cities, organizations, and named individuals.
2. It also identifies many relevant solar-market concepts.
3. The four generic entity types are insufficient for the document domain.
4. Many technologies, policies, targets, infrastructure elements, market concepts, and stakeholder groups are forced into `ORGANIZATION`, `EVENT`, `PERSON`, or `GEO`.
5. Abbreviations, shortened names, plural variants, and reordered person names are not reliably resolved.
6. LLM-generated descriptions may extend beyond what the source states explicitly.

The largest apparent problem at this stage is therefore not a complete failure to recognize concepts. It is the combination of:

* **Entity-type/schema failure**
* **Entity-resolution failure**
* **Description overgeneralization**

---

## Current status and limitation

This section reports preliminary patterns identified through inspection of the complete output and the reproducible 57-entity sample.

The individual sample rows have not yet received final source-based classifications. Therefore, no final error percentages are reported at this point.

The next step is to trace the sampled entities to their supporting `text_unit_ids` and classify them as:

* `CORRECT_AND_RELEVANT`
* `CORRECT_BUT_IRRELEVANT`
* `WRONG_TYPE`
* `DUPLICATE_OR_ALIAS`
* `OVERGENERALIZED`
* `UNSUPPORTED_OR_HALLUCINATED`

The completed classification will support approximate error proportions and a final assessment of entity-extraction quality.


In [6]:
def add_review(
    title,
    source_supported,
    classification,
    expected_type,
    duplicate_or_alias_of="",
    description_problem="",
    pipeline_stage="",
    notes="",
):
    mask = entity_review["title"] == title

    entity_review.loc[mask, "source_supported"] = source_supported
    entity_review.loc[mask, "primary_classification"] = classification
    entity_review.loc[mask, "expected_type"] = expected_type
    entity_review.loc[mask, "duplicate_or_alias_of"] = duplicate_or_alias_of
    entity_review.loc[mask, "description_problem"] = description_problem
    entity_review.loc[mask, "likely_pipeline_stage"] = pipeline_stage
    entity_review.loc[mask, "audit_notes"] = notes

In [7]:
add_review(
    title="ÖSTERREICH",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="COUNTRY",
    description_problem="Description contains broad claims not all explicitly stated in one source passage.",
    pipeline_stage="DESCRIPTION_SUMMARIZATION",
    notes="Austria is explicitly and repeatedly mentioned in the source."
)

add_review(
    title="ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="POLICY",
    description_problem="Description combines and expands several themes from across the document.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="The strategy is a policy/strategy document, not an event."
)

add_review(
    title="PV-ANLAGE",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="TECHNOLOGY_OR_INSTALLATION",
    duplicate_or_alias_of="PV-ANLAGEN / PHOTOVOLTAIKANLAGEN",
    description_problem="Description generalizes PV installations as decentralized systems.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION",
    notes="A PV installation is not an organization."
)

add_review(
    title="BMK",
    source_supported="YES",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="ORGANIZATION",
    duplicate_or_alias_of="BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENERGIE, MOBILITÄT, INNOVATION UND TECHNOLOGIE",
    description_problem="Description expands the ministry's general responsibilities.",
    pipeline_stage="ENTITY_RESOLUTION_AND_DESCRIPTION_SUMMARIZATION",
    notes="BMK is the abbreviation of the full ministry name and should map to one canonical organization."
)

add_review(
    title="KLIMA- UND ENERGIEFONDS",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    pipeline_stage="NONE",
    notes="The Climate and Energy Fund is a genuine organization relevant to Austrian energy policy."
)

In [10]:
add_review(
    title="BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENERGIE, MOBILITÄT, INNOVATION UND TECHNOLOGIE",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    duplicate_or_alias_of="Canonical entity; BMK should map to this entity.",
    description_problem="Description expands the ministry's responsibilities beyond the immediate publication information.",
    pipeline_stage="DESCRIPTION_SUMMARIZATION",
    notes="The ministry is explicitly identified as publisher of the strategy."
)

add_review(
    title="EUROPÄISCHE KOMMISSION",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    pipeline_stage="NONE",
    notes="The European Commission is a named institutional actor in the policy context."
)

add_review(
    title="WIEN",
    source_supported="YES",
    classification="CORRECT_BUT_IRRELEVANT",
    expected_type="CITY",
    pipeline_stage="RELEVANCE_FILTERING",
    notes="Vienna appears mainly as publication and ministry-address information and adds little to the intended solar-market KG."
)

add_review(
    title="EU",
    source_supported="YES",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="ORGANIZATION",
    duplicate_or_alias_of="EUROPÄISCHE UNION",
    description_problem="Description summarizes the EU's broad political role.",
    pipeline_stage="ENTITY_RESOLUTION",
    notes="EU is an abbreviation and should map to the canonical European Union entity."
)

add_review(
    title="PV-ANLAGEN",
    source_supported="YES",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="TECHNOLOGY_OR_INSTALLATION",
    duplicate_or_alias_of="PV-ANLAGE / PHOTOVOLTAIKANLAGEN",
    description_problem="The description interprets installations as an event involving deployment and operation.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION",
    notes="Plural form of PV installation; incorrectly typed as EVENT."
)

add_review(
    title="BUNDESVERBAND PV AUSTRIA",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    pipeline_stage="NONE",
    notes="PV Austria is a relevant photovoltaic-sector association explicitly referenced in the document."
)

add_review(
    title="ENERGIEWENDE",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="PROCESS",
    description_problem="Description treats a broad transformation process as a single bounded initiative.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_OVERGENERALIZATION",
    notes="The energy transition is a relevant process, not a discrete event."
)

add_review(
    title="INTEGRIERTER ÖSTERREICHISCHER NETZINFRASTRUKTURPLAN (NIP)",
    source_supported="PARTIAL",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="POLICY_OR_PLAN",
    duplicate_or_alias_of="INTEGRIERTER ÖSTERREICHISCHER NETZINFRASTRUKTURPLAN (ÖNIP)",
    description_problem="The abbreviation NIP is inconsistent with the document's principal Austrian abbreviation ÖNIP.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION",
    notes="The plan is real, but it should be normalized to the canonical ÖNIP entity and should not be typed as an organization."
)

add_review(
    title="PHOTOVOLTAIK-STRATEGIE",
    source_supported="YES",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="POLICY",
    duplicate_or_alias_of="ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION",
    notes="Shortened name for the same Austrian strategy; incorrectly maintained as a separate EVENT."
)

add_review(
    title="STROMNETZ",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="INFRASTRUCTURE",
    description_problem="Description correctly identifies an electricity grid but assigns it organizational characteristics.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="The electricity grid is physical/system infrastructure, not an organization."
)

In [11]:
entity_review[
    entity_review["primary_classification"] != ""
][
    [
        "title",
        "type",
        "source_supported",
        "primary_classification",
        "expected_type",
        "duplicate_or_alias_of",
        "likely_pipeline_stage",
        "audit_notes",
    ]
]

,title,type,source_supported,primary_classification,expected_type,duplicate_or_alias_of,likely_pipeline_stage,audit_notes
0,ÖSTERREICH,GEO,YES,CORRECT_AND_RELEVANT,COUNTRY,,DESCRIPTION_SUMMARIZATION,Austria is explicitly and repeatedly mentioned in the source.
1,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,EVENT,YES,WRONG_TYPE,POLICY,,ENTITY_TYPE_SCHEMA,"The strategy is a policy/strategy document, not an event."
2,PV-ANLAGE,ORGANIZATION,YES,WRONG_TYPE,TECHNOLOGY_OR_INSTALLATION,PV-ANLAGEN / PHOTOVOLTAIKANLAGEN,ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION,A PV installation is not an organization.
3,BMK,ORGANIZATION,YES,DUPLICATE_OR_ALIAS,ORGANIZATION,"BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENERGIE, MOBILITÄT, INNOVATION UND TECHNOLOGIE",ENTITY_RESOLUTION_AND_DESCRIPTION_SUMMARIZATION,BMK is the abbreviation of the full ministry name and should map to one canonical organization.
4,KLIMA- UND ENERGIEFONDS,ORGANIZATION,YES,CORRECT_AND_RELEVANT,ORGANIZATION,,NONE,The Climate and Energy Fund is a genuine organization relevant to Austrian energy policy.
5,ENERGIEGEMEINSCHAFTEN,ORGANIZATION,YES,CORRECT_AND_RELEVANT,ORGANIZATION_OR_MARKET_ACTOR,,DESCRIPTION_SUMMARIZATION,Energy communities are explicitly discussed as important participation mechanisms.
6,BUND,ORGANIZATION,YES,OVERGENERALIZED,GOVERNMENT_LEVEL_OR_ACTOR,,ENTITY_IDENTIFICATION_AND_SCHEMA,"Relevant concept, but represented as an overly broad organization."
7,PHOTOVOLTAIKANLAGEN,ORGANIZATION,YES,DUPLICATE_OR_ALIAS,TECHNOLOGY_OR_INSTALLATION,PV-ANLAGE / PV-ANLAGEN,ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION,Plural German form of photovoltaic installations; not an organization.
8,E-CONTROL,ORGANIZATION,YES,CORRECT_AND_RELEVANT,ORGANIZATION,,NONE,E-Control is an explicitly mentioned Austrian regulatory authority.
9,BUNDESLÄNDER,GEO,YES,CORRECT_AND_RELEVANT,ADMINISTRATIVE_REGION,,MINOR_SCHEMA_LIMITATION,The Austrian federal states are relevant geographic and administrative entities.


In [12]:
add_review(
    title="ENERGIEGEMEINSCHAFTEN",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION_OR_MARKET_ACTOR",
    description_problem="Description may generalize different legal and organizational forms.",
    pipeline_stage="DESCRIPTION_SUMMARIZATION",
    notes="Energy communities are explicitly discussed as important participation mechanisms."
)

add_review(
    title="BUND",
    source_supported="YES",
    classification="OVERGENERALIZED",
    expected_type="GOVERNMENT_LEVEL_OR_ACTOR",
    description_problem="The term Bund refers broadly to the federal level rather than one clearly bounded organization.",
    pipeline_stage="ENTITY_IDENTIFICATION_AND_SCHEMA",
    notes="Relevant concept, but represented as an overly broad organization."
)

add_review(
    title="PHOTOVOLTAIKANLAGEN",
    source_supported="YES",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="TECHNOLOGY_OR_INSTALLATION",
    duplicate_or_alias_of="PV-ANLAGE / PV-ANLAGEN",
    description_problem="Description incorrectly frames installations partly as organizations or actors.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION",
    notes="Plural German form of photovoltaic installations; not an organization."
)

add_review(
    title="E-CONTROL",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    pipeline_stage="NONE",
    notes="E-Control is an explicitly mentioned Austrian regulatory authority."
)

add_review(
    title="BUNDESLÄNDER",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ADMINISTRATIVE_REGION",
    description_problem="Description summarizes the collective role of multiple federal states.",
    pipeline_stage="MINOR_SCHEMA_LIMITATION",
    notes="The Austrian federal states are relevant geographic and administrative entities."
)

In [13]:
add_review(
    title="VERORDNUNGEN",
    source_supported="YES",
    classification="OVERGENERALIZED",
    expected_type="LAW_OR_REGULATION_COLLECTION",
    description_problem="The plural term groups multiple regulations into one vague entity.",
    pipeline_stage="ENTITY_IDENTIFICATION_AND_SCHEMA",
    notes="Regulations are relevant, but specific legal instruments should normally be represented separately."
)

add_review(
    title="STROMMARKTPREISE",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="MARKET_METRIC",
    description_problem="Description summarizes market-price behaviour rather than identifying a discrete event.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Electricity-market prices are a market indicator, not an event."
)

add_review(
    title="EU-TAXONOMY",
    source_supported="PARTIAL",
    classification="WRONG_TYPE",
    expected_type="POLICY_OR_REGULATION",
    description_problem="Entity label is an English normalization of the German source reference, while the description adds a general explanation.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_DESCRIPTION_SUMMARIZATION",
    notes="The EU taxonomy is a regulatory classification framework, not an event."
)

add_review(
    title="BIODIVERSITÄT",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="ENVIRONMENTAL_FACTOR",
    description_problem="Description treats a broad environmental concept as an event or outcome.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Biodiversity is a relevant environmental factor, not an event."
)

add_review(
    title="MADE IN AUSTRIA",
    source_supported="YES",
    classification="OVERGENERALIZED",
    expected_type="INDUSTRIAL_LABEL_OR_STRATEGIC_CONCEPT",
    description_problem="A phrase describing Austrian production is promoted into a standalone entity.",
    pipeline_stage="ENTITY_IDENTIFICATION_AND_SCHEMA",
    notes="Relevant to domestic value creation, but not a discrete event."
)

add_review(
    title="STRATEGIEPLAN FÜR ENERGIETECHNOLOGIE (SET-PLAN)",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="POLICY_OR_PLAN",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="The SET Plan is a European strategic policy plan, not an event."
)

add_review(
    title="REGIERUNGSPROGRAMM",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="POLICY",
    description_problem="Description represents the government programme broadly without identifying a precise version or period.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_ENTITY_SPECIFICITY",
    notes="A government programme is a policy document, not an event."
)

add_review(
    title="AKTIONSPLAN NETZANSCHLUSS 2023",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="POLICY_OR_ACTION_PLAN",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="The Grid Connection Action Plan 2023 is a named action plan, not an event."
)

add_review(
    title="RICHTLINIE 98/70/EG",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="LAW_OR_DIRECTIVE",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="An EU directive is a legal instrument, not an event."
)

add_review(
    title="GEBÄUDE DES BUNDES",
    source_supported="PARTIAL",
    classification="WRONG_TYPE",
    expected_type="BUILDING_CATEGORY_OR_INFRASTRUCTURE",
    description_problem="GraphRAG converts a general reference to federally owned or used buildings into a geographic entity.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_OVERGENERALIZATION",
    notes="Federal buildings are physical infrastructure or a building category, not a geographic location."
)

In [14]:
add_review(
    title="STRASSEN- UND BAHNBEREICH",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="INFRASTRUCTURE_OR_LAND_USE",
    description_problem="Description treats transport-infrastructure areas as geographic entities.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Road and railway areas are infrastructure or potential PV site categories, not conventional geographic entities."
)

add_review(
    title="MITGLIEDSTAATEN",
    source_supported="YES",
    classification="OVERGENERALIZED",
    expected_type="COUNTRY_GROUP",
    description_problem="The plural term combines unspecified EU countries into one entity.",
    pipeline_stage="ENTITY_IDENTIFICATION_AND_SPECIFICITY",
    notes="EU member states are relevant collectively, but this is not a clearly bounded individual geographic entity."
)

add_review(
    title="ERNNEUERBAREN-BESCHLEUNIGUNGSGEBIETE (RENEWABLES ACCELERATION AREAS)",
    source_supported="PARTIAL",
    classification="CORRECT_AND_RELEVANT",
    expected_type="REGULATORY_ZONE",
    description_problem="The entity contains an apparent spelling error and an English translation added to the German label.",
    pipeline_stage="LABEL_NORMALIZATION_AND_MINOR_SCHEMA_LIMITATION",
    notes="Renewables acceleration areas are relevant policy-designated zones, although the label requires normalization."
)

add_review(
    title="BETRIEBSANLAGEN",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="INFRASTRUCTURE_OR_FACILITY",
    description_problem="Description interprets operational facilities as geographic sites.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Operational facilities are physical facilities or a site category, not a GEO entity."
)

add_review(
    title="REIHENHÄUSER",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="BUILDING_TYPE",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Terraced houses are a building type, not a geographic entity."
)

add_review(
    title="EINZELGEBÄUDE",
    source_supported="PARTIAL",
    classification="WRONG_TYPE",
    expected_type="BUILDING_TYPE",
    description_problem="The entity normalizes a general reference to individual buildings into a separate GEO entity.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_NORMALIZATION",
    notes="An individual building is a physical object or building category, not a geographic entity."
)

add_review(
    title="NATUR",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="ENVIRONMENTAL_FACTOR",
    description_problem="A broad environmental concept is represented as a geographic entity.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_OVERGENERALIZATION",
    notes="Nature is relevant to environmental protection and acceptance, but it is not a GEO entity."
)

add_review(
    title="FREIFLÄCHEN",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="LAND_USE_OR_SITE_TYPE",
    description_problem="The plural term refers broadly to open-land areas rather than a specific place.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Open land is a PV site or land-use category, not a specific geographic entity."
)

add_review(
    title="EU SOLAR PV INDUSTRY ALLIANCE",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    pipeline_stage="NONE",
    notes="The alliance is a named European organization relevant to strengthening the PV industry."
)

add_review(
    title="EU-EMISSIONSHANDEL",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="POLICY_OR_MARKET_MECHANISM",
    description_problem="The emissions-trading system is represented as an organization.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="EU emissions trading is a regulatory market mechanism, not an organization."
)

In [16]:
entity_review[
    entity_review["primary_classification"] != ""
]

,id,title,type,description,text_unit_ids,frequency,degree,source_supported,primary_classification,expected_type,duplicate_or_alias_of,description_problem,likely_pipeline_stage,audit_notes
0,05c9af58-ac32-4ec6-b923-358b587e8cae,ÖSTERREICH,GEO,"Österreich, also known as Austria, is a Central European nation and a member of the European Union that is at the fo...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...,15,82,YES,CORRECT_AND_RELEVANT,COUNTRY,,Description contains broad claims not all explicitly stated in one source passage.,DESCRIPTION_SUMMARIZATION,Austria is explicitly and repeatedly mentioned in the source.
1,cb0a1b84-f9a3-4404-9807-afa3d1e6c3af,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,EVENT,"The ""Österreichische Photovoltaik-Strategie"" (Austrian Photovoltaic Strategy) is Austria’s official national strateg...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...,14,39,YES,WRONG_TYPE,POLICY,,Description combines and expands several themes from across the document.,ENTITY_TYPE_SCHEMA,"The strategy is a policy/strategy document, not an event."
2,87671c51-7ba4-4fb4-97a0-6e4dde071aaf,PV-ANLAGE,ORGANIZATION,PV-ANLAGE (Photovoltaic system) refers to decentralized solar power generation installations that are widely deploye...,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...,3,20,YES,WRONG_TYPE,TECHNOLOGY_OR_INSTALLATION,PV-ANLAGEN / PHOTOVOLTAIKANLAGEN,Description generalizes PV installations as decentralized systems.,ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION,A PV installation is not an organization.
3,a904e0ff-8af4-4090-9452-b6ee49f26cb8,BMK,ORGANIZATION,"BMK, the commonly used abbreviation for the Austrian Federal Ministry for Climate Action, Environment, Energy, Mobil...",['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...,7,16,YES,DUPLICATE_OR_ALIAS,ORGANIZATION,"BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENERGIE, MOBILITÄT, INNOVATION UND TECHNOLOGIE",Description expands the ministry's general responsibilities.,ENTITY_RESOLUTION_AND_DESCRIPTION_SUMMARIZATION,BMK is the abbreviation of the full ministry name and should map to one canonical organization.
4,3f8fbdde-b872-4a1e-b5d9-dd3c12a82852,KLIMA- UND ENERGIEFONDS,ORGANIZATION,The **KLIMA- UND ENERGIEFONDS** (Climate and Energy Fund) is an Austrian governmental and public organization dedica...,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...,6,14,YES,CORRECT_AND_RELEVANT,ORGANIZATION,,,NONE,The Climate and Energy Fund is a genuine organization relevant to Austrian energy policy.
5,28ad2239-3425-467c-a5b7-d2d823e3d34d,ENERGIEGEMEINSCHAFTEN,ORGANIZATION,ENERGIEGEMEINSCHAFTEN (energy communities) are formal or informal organizations or associations composed of citizens...,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...,6,14,YES,CORRECT_AND_RELEVANT,ORGANIZATION_OR_MARKET_ACTOR,,Description may generalize different legal and organizational forms.,DESCRIPTION_SUMMARIZATION,Energy communities are explicitly discussed as important participation mechanisms.
6,9146b3ae-f272-4a61-beba-e5b1773d1520,BUND,ORGANIZATION,"The federal government of Austria, which collaborates with the states (Bundesländer) to achieve national renewable e...",['23663988970639bea940978ccbd9150fd292856f2f31a1d587392606a64713d5dfeade66f1beea67d1cc0b0e93012337685e0eca3267f98c16...,1,14,YES,OVERGENERALIZED,GOVERNMENT_LEVEL_OR_ACTOR,,The term Bund refers broadly to the federal level rather than one clearly bounded organization.,ENTITY_IDENTIFICATION_AND_SCHEMA,"Relevant concept, but represented as an overly broad organization."
7,9983cf95-be92-4fe5-afab-1ec698234a7d,PHOTOVOLTAIKANLAGEN,ORGANIZATIO

In [17]:
add_review(
    title="BIOENERGIE",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="TECHNOLOGY_OR_ENERGY_SOURCE",
    description_problem="Bioenergy is described as a sector but classified as an organization.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Bioenergy is an energy source or technology field, not an organization."
)

add_review(
    title="ENERGIEBERATUNGSSTELLEN DER BUNDESLÄNDER",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    description_problem="The plural label groups several regional advisory bodies into one collective entity.",
    pipeline_stage="MINOR_ENTITY_RESOLUTION",
    notes="These are relevant energy-advisory organizations, although individual regional bodies are not distinguished."
)

add_review(
    title="WASSERSTOFF",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="ENERGY_CARRIER_OR_TECHNOLOGY",
    description_problem="Hydrogen is described as a technological field but classified as an organization.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Hydrogen is an energy carrier or technology, not an organization."
)

add_review(
    title="VERKEHRSSEKTOR",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="SECTOR",
    description_problem="An economic sector is represented as an organization.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="The transport sector is a sector, not one organization."
)

add_review(
    title="EUROPEAN ECONOMIC AND SOCIAL COMMITTEE",
    source_supported="YES",
    classification="CORRECT_BUT_IRRELEVANT",
    expected_type="ORGANIZATION",
    pipeline_stage="RELEVANCE_FILTERING",
    notes="The committee is a real institution, but it appears mainly in reference or wider EU-policy context."
)

add_review(
    title="EUROPAPARLAMENT",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="ORGANIZATION",
    pipeline_stage="NONE",
    notes="The European Parliament is a relevant institutional actor in the EU regulatory context."
)

add_review(
    title="PV",
    source_supported="YES",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="TECHNOLOGY",
    duplicate_or_alias_of="PHOTOVOLTAIK",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_ENTITY_RESOLUTION",
    notes="PV is the common abbreviation for photovoltaics and should map to one canonical technology entity."
)

add_review(
    title="BEVÖLKERUNG",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="STAKEHOLDER_GROUP",
    description_problem="The entire population is represented as if it were an individual person.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_OVERGENERALIZATION",
    notes="The population is a collective stakeholder group, not a PERSON entity."
)

add_review(
    title="DACHDECKER:INNEN",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="STAKEHOLDER_GROUP_OR_OCCUPATION",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Roofers are an occupational stakeholder group, not one person."
)

add_review(
    title="LANDESENERGIEREFERENT:INNEN",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="STAKEHOLDER_GROUP_OR_GOVERNMENT_ROLE",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Regional energy representatives are a group of officials or roles, not one person."
)

add_review(
    title="HACKNER, JOHANNES",
    source_supported="YES",
    classification="CORRECT_BUT_IRRELEVANT",
    expected_type="PERSON",
    duplicate_or_alias_of="Canonical source form; JOHANNES HACKNER should map to this entity.",
    description_problem="Description presents the cited author as a broader renewable-energy expert.",
    pipeline_stage="DESCRIPTION_SUMMARIZATION_AND_RELEVANCE_FILTERING",
    notes="The person appears primarily as an author in a citation and is not a central solar-market entity."
)

add_review(
    title="TEILNEHMER:INNEN MEHRFAMILIENHÄUSER",
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    expected_type="STAKEHOLDER_GROUP",
    description_problem="GraphRAG combines participants and multifamily-building context into an artificial entity label.",
    pipeline_stage="ENTITY_IDENTIFICATION_AND_SCHEMA",
    notes="The concept refers broadly to participants or residents in multifamily buildings, not a named person."
)

add_review(
    title="JOHANNES HACKNER",
    source_supported="PARTIAL",
    classification="DUPLICATE_OR_ALIAS",
    expected_type="PERSON",
    duplicate_or_alias_of="HACKNER, JOHANNES",
    pipeline_stage="ENTITY_RESOLUTION",
    notes="Reordered version of the same cited person's name."
)

add_review(
    title="LEONORE GEWESSLER",
    source_supported="YES",
    classification="CORRECT_AND_RELEVANT",
    expected_type="PERSON",
    pipeline_stage="NONE",
    notes="Leonore Gewessler is explicitly named as the minister and author of the foreword."
)

add_review(
    title="AKTIVER KUNDE",
    source_supported="YES",
    classification="WRONG_TYPE",
    expected_type="MARKET_ACTOR_OR_LEGAL_ROLE",
    description_problem="A legal market-participant category is treated as an individual person.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Active customer is a defined legal or market role, not one named person."
)

add_review(
    title="PRIVATE INVESTOREN",
    source_supported="PARTIAL",
    classification="WRONG_TYPE",
    expected_type="STAKEHOLDER_GROUP",
    description_problem="A general investor category is represented as one person entity.",
    pipeline_stage="ENTITY_TYPE_SCHEMA_AND_NORMALIZATION",
    notes="Private investors are a stakeholder group, not an individual person."
)

add_review(
    title="PRIVATPERSONEN",
    source_supported="PARTIAL",
    classification="WRONG_TYPE",
    expected_type="STAKEHOLDER_GROUP",
    description_problem="A collective category of private individuals is represented as one person entity.",
    pipeline_stage="ENTITY_TYPE_SCHEMA",
    notes="Private individuals form a stakeholder group rather than one named person."
)

In [18]:
reviewed = entity_review[
    entity_review["primary_classification"] != ""
]

print("Reviewed entities:", len(reviewed))
print()
print("Primary classifications:")
print(reviewed["primary_classification"].value_counts())
print()
print("Source support:")
print(reviewed["source_supported"].value_counts())

Reviewed entities: 57

Primary classifications:
primary_classification
WRONG_TYPE                28
CORRECT_AND_RELEVANT      13
DUPLICATE_OR_ALIAS         8
OVERGENERALIZED            5
CORRECT_BUT_IRRELEVANT     3
Name: count, dtype: int64

Source support:
source_supported
YES        48
PARTIAL     9
Name: count, dtype: int64


## Stage 3 — Completed Entity-Extraction Audit

### Audit scope

A reproducible sample of **57 entities** was reviewed from the complete GraphRAG output of 312 entities.

The sample combined:

* The 20 most highly connected entities
* Ten randomly selected entities from each raw GraphRAG type
* Removal of duplicated rows appearing in both groups
* A fixed random seed of `42`

Each sampled entity was evaluated according to:

* Source support
* Relevance to the intended solar-market KG
* Entity-type correctness
* Duplicate or alias status
* Degree of generalization
* Description quality
* Likely pipeline stage responsible for any problem

The original Austrian Photovoltaic Strategy and the associated text units were treated as the authoritative evidence.

---

## Classification results

All 57 sampled entities received one primary classification.

| Primary classification        |  Count | Approximate share |
| ----------------------------- | -----: | ----------------: |
| `WRONG_TYPE`                  |     28 |             49.1% |
| `CORRECT_AND_RELEVANT`        |     13 |             22.8% |
| `DUPLICATE_OR_ALIAS`          |      8 |             14.0% |
| `OVERGENERALIZED`             |      5 |              8.8% |
| `CORRECT_BUT_IRRELEVANT`      |      3 |              5.3% |
| `UNSUPPORTED_OR_HALLUCINATED` |      0 |              0.0% |
| **Total**                     | **57** |          **100%** |

Only **13 of 57 sampled entities**, or approximately **22.8%**, were classified as both correct and relevant without a more important primary problem.

The most common primary error was incorrect entity typing. Nearly half of the sample was assigned an unsuitable GraphRAG type.

---

## Source-support results

| Source-support assessment |  Count | Approximate share |
| ------------------------- | -----: | ----------------: |
| `YES`                     |     48 |             84.2% |
| `PARTIAL`                 |      9 |             15.8% |
| `NO`                      |      0 |              0.0% |
| **Total**                 | **57** |          **100%** |

The high source-support rate indicates that GraphRAG usually extracted concepts related to text found in the document.

However, source support does not mean that the resulting entity was correctly represented. A concept can be present in the source but still be:

* Assigned the wrong type
* Given an artificial or normalized label
* Duplicated under another name
* Overgeneralized
* Described using unsupported language
* Irrelevant to the intended KG

The results therefore suggest that **entity identification is considerably better than entity typing and entity resolution**.

---

## Results by raw GraphRAG type

| Raw type       | Correct and relevant | Correct but irrelevant | Duplicate/alias | Overgeneralized | Wrong type |  Total |
| -------------- | -------------------: | ---------------------: | --------------: | --------------: | ---------: | -----: |
| `EVENT`        |                    0 |                      0 |               2 |               2 |          9 |     13 |
| `GEO`          |                    3 |                      1 |               0 |               1 |          7 |     12 |
| `ORGANIZATION` |                    9 |                      1 |               5 |               1 |          6 |     22 |
| `PERSON`       |                    1 |                      1 |               1 |               1 |          6 |     10 |
| **Total**      |               **13** |                  **3** |           **8** |           **5** |     **28** | **57** |

### Interpretation

The `EVENT` category performed especially poorly. None of the 13 sampled `EVENT` entities was classified as both correct and relevant under that type.

The `EVENT` category absorbed concepts such as:

* Policies
* Strategies
* Laws
* Market metrics
* Environmental factors
* Processes
* Action plans
* Industrial labels

The `ORGANIZATION` category performed better for conventional named institutions, but it also absorbed technologies, infrastructure, sectors, and market mechanisms.

The `PERSON` category incorrectly included collective stakeholder groups and legal market roles.

The `GEO` category incorrectly included building types, infrastructure, land-use categories, and environmental concepts.

---

## Entity-identification quality

GraphRAG generally succeeded in identifying concepts that were connected to the source document.

Examples of successfully identified entities include:

* Austria
* E-Control
* Austrian Ministry for Climate Action
* Climate and Energy Fund
* European Commission
* PV Austria
* European Parliament
* Energy communities
* Leonore Gewessler
* EU Solar PV Industry Alliance

This indicates that the extraction model can recognize conventional named entities and many important domain concepts.

The principal problem is not that GraphRAG extracted completely unrelated words in every case. Instead, it often converted meaningful source concepts into poorly modelled graph entities.

---

## Entity-type and schema problems

The GraphRAG configuration allowed only:

```yaml
entity_types:
  - organization
  - person
  - geo
  - event
```

These four categories are insufficient for an Austrian solar-market policy document.

### Policies and plans classified as events

| GraphRAG output                  | Raw type | Better type               |
| -------------------------------- | -------- | ------------------------- |
| Austrian Photovoltaic Strategy   | `EVENT`  | `POLICY` or `STRATEGY`    |
| SET Plan                         | `EVENT`  | `POLICY` or `PLAN`        |
| Government programme             | `EVENT`  | `POLICY`                  |
| Grid Connection Action Plan 2023 | `EVENT`  | `POLICY` or `ACTION_PLAN` |
| Directive 98/70/EC               | `EVENT`  | `LAW` or `DIRECTIVE`      |

### Technologies and infrastructure classified as organizations

| GraphRAG output            | Raw type       | Better type                      |
| -------------------------- | -------------- | -------------------------------- |
| PV installation            | `ORGANIZATION` | `TECHNOLOGY` or `INSTALLATION`   |
| Photovoltaic installations | `ORGANIZATION` | `TECHNOLOGY` or `INSTALLATION`   |
| Electricity grid           | `ORGANIZATION` | `INFRASTRUCTURE`                 |
| Bioenergy                  | `ORGANIZATION` | `TECHNOLOGY` or `ENERGY_SOURCE`  |
| Hydrogen                   | `ORGANIZATION` | `ENERGY_CARRIER` or `TECHNOLOGY` |

### Market and regulatory concepts assigned unsuitable types

| GraphRAG output           | Raw type       | Better type                    |
| ------------------------- | -------------- | ------------------------------ |
| Electricity-market prices | `EVENT`        | `MARKET_METRIC`                |
| EU emissions trading      | `ORGANIZATION` | `POLICY` or `MARKET_MECHANISM` |
| EU taxonomy               | `EVENT`        | `POLICY` or `REGULATION`       |
| Active customer           | `PERSON`       | `MARKET_ACTOR` or `LEGAL_ROLE` |

### Physical and land-use concepts classified as geography

| GraphRAG output        | Raw type | Better type                             |
| ---------------------- | -------- | --------------------------------------- |
| Federal buildings      | `GEO`    | `INFRASTRUCTURE` or `BUILDING_CATEGORY` |
| Road and railway areas | `GEO`    | `INFRASTRUCTURE` or `LAND_USE`          |
| Operational facilities | `GEO`    | `FACILITY` or `INFRASTRUCTURE`          |
| Terraced houses        | `GEO`    | `BUILDING_TYPE`                         |
| Open land              | `GEO`    | `LAND_USE` or `SITE_TYPE`               |

### Stakeholder groups classified as persons

| GraphRAG output                 | Raw type | Better type                              |
| ------------------------------- | -------- | ---------------------------------------- |
| Population                      | `PERSON` | `STAKEHOLDER_GROUP`                      |
| Roofers                         | `PERSON` | `STAKEHOLDER_GROUP` or `OCCUPATION`      |
| Regional energy representatives | `PERSON` | `STAKEHOLDER_GROUP` or `GOVERNMENT_ROLE` |
| Private investors               | `PERSON` | `STAKEHOLDER_GROUP`                      |
| Private individuals             | `PERSON` | `STAKEHOLDER_GROUP`                      |

These results confirm that entity typing is a **major pipeline bottleneck**.

---

## Duplicate and alias problems

Eight sampled entities received `DUPLICATE_OR_ALIAS` as their primary classification.

### Ministry abbreviation

```text
BMK
```

and:

```text
BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENERGIE,
MOBILITÄT, INNOVATION UND TECHNOLOGIE
```

refer to the same organization.

The full ministry name should function as the canonical entity, with `BMK` stored as an alias.

### Austrian Photovoltaic Strategy

GraphRAG generated variants including:

```text
ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE
PHOTOVOLTAIK-STRATEGIE
PV-STRATEGIE
```

These should generally map to one canonical policy entity.

### Photovoltaic installations

GraphRAG generated variants including:

```text
PV-ANLAGE
PV-ANLAGEN
PHOTOVOLTAIKANLAGEN
```

Some variants also received different raw types. This demonstrates an interaction between typing and entity resolution.

### European Union

```text
EU
```

should map to a canonical European Union entity.

### Photovoltaics

```text
PV
```

should normally map to the canonical technology concept `Photovoltaics`, while preserving `PV` as an abbreviation.

### Austrian network infrastructure plan

```text
INTEGRIERTER ÖSTERREICHISCHER NETZINFRASTRUKTURPLAN (NIP)
```

appears to be a variant of:

```text
INTEGRIERTER ÖSTERREICHISCHER NETZINFRASTRUKTURPLAN (ÖNIP)
```

The source principally uses the Austrian abbreviation `ÖNIP`. The inconsistent `NIP` form requires normalization.

### Personal name order

```text
HACKNER, JOHANNES
```

and:

```text
JOHANNES HACKNER
```

refer to the same person.

These cases show that adding domain-specific entity types will not be sufficient by itself. A later entity-resolution step is still required.

---

## Overgeneralized entities

Five sampled entities were classified primarily as `OVERGENERALIZED`.

Examples include:

* `BUND`
* `VERORDNUNGEN`
* `MADE IN AUSTRIA`
* `MITGLIEDSTAATEN`
* `TEILNEHMER:INNEN MEHRFAMILIENHÄUSER`

These entities often originate from real source language but lack clear boundaries.

For example:

* `VERORDNUNGEN` combines unspecified regulations rather than representing individual legal instruments.
* `BUND` refers broadly to the federal level rather than one precisely defined organization.
* `MITGLIEDSTAATEN` groups unspecified EU countries.
* `TEILNEHMER:INNEN MEHRFAMILIENHÄUSER` appears to combine a stakeholder group with a building context into an artificial entity label.

Overgeneralization reduces explainability because it becomes unclear what exact real-world entity the node represents.

---

## Correct but irrelevant entities

Three sampled entities were classified as correct but insufficiently relevant to the intended solar-market KG:

* Vienna
* European Economic and Social Committee
* Johannes Hackner in citation context

For example, Vienna appears mainly as publication or ministry-address information. It is a real geographic entity, but its presence does not necessarily help explain Austrian solar-market policy, market, or technical relationships.

Similarly, an author appearing only in a bibliographic citation may not need to become a central domain entity.

This finding supports filtering document front matter, contact information, and references before domain extraction.

---

## Description-quality problems

The `description` field is generated by an LLM. It is not copied directly from the source.

Several descriptions use broad language such as:

* “at the forefront”
* “plays a crucial role”
* “leading role”
* “widely deployed”
* “central hub”
* “comprehensive initiative”

These phrases may combine information from multiple chunks or introduce interpretation.

For example:

```text
Source evidence:
The document identifies the ministry as publisher and discusses policies in its area.

GraphRAG description:
The ministry is described as the principal driver responsible for developing,
executing, and overseeing a broad range of national strategies.
```

The description may be broadly plausible, but not every part is necessarily stated explicitly in the supporting text.

Description quality must therefore be audited independently from entity-name and entity-type correctness.

---

## Representative error traces

| Source evidence                                                               | GraphRAG output                                  | Problem                                | Likely stage          | Recommended fix                                       |
| ----------------------------------------------------------------------------- | ------------------------------------------------ | -------------------------------------- | --------------------- | ----------------------------------------------------- |
| The source is an Austrian PV strategy document                                | Strategy → `EVENT`                               | Wrong type                             | Entity schema         | Add `POLICY`/`STRATEGY`                               |
| PV installations are technical systems                                        | PV installation → `ORGANIZATION`                 | Wrong type                             | Entity schema         | Add installation/technology type                      |
| The electricity grid is infrastructure                                        | Electricity grid → `ORGANIZATION`                | Wrong type                             | Entity schema         | Add `INFRASTRUCTURE`                                  |
| Electricity-market prices describe market conditions                          | Prices → `EVENT`                                 | Wrong type                             | Entity schema         | Add `MARKET_METRIC`                                   |
| BMK is the ministry abbreviation                                              | BMK and full ministry name become separate nodes | Alias duplication                      | Entity resolution     | Canonical-name dictionary                             |
| PV, PV installation, and photovoltaic installations appear in different forms | Several nodes with differing types               | Alias and granularity conflict         | Typing and resolution | Define canonical concepts and aliases                 |
| Roofers are a professional stakeholder group                                  | Roofers → `PERSON`                               | Collective group treated as individual | Entity schema         | Add `STAKEHOLDER_GROUP`                               |
| Contact and publication information identifies Vienna                         | Vienna becomes graph entity                      | Correct but low relevance              | Input filtering       | Exclude or separately process administrative material |

---

## Missing domain categories

The results strongly support expanding the extraction schema beyond the four generic types.

Candidate categories supported by this source include:

* `COUNTRY`
* `ORGANIZATION`
* `POLICY`
* `LAW`
* `SUPPORT_SCHEME`
* `TECHNOLOGY`
* `INFRASTRUCTURE`
* `TARGET`
* `MARKET_METRIC`
* `CONSTRAINT`
* `STAKEHOLDER_GROUP`
* `PROCESS`
* `LAND_USE`
* `BUILDING_TYPE`
* `ENVIRONMENTAL_FACTOR`
* `MARKET_ACTOR`

This does not mean that every candidate must automatically become a final Schema V2 class. The final controlled schema should balance domain coverage with simplicity.

---

## Pipeline diagnosis

### Entity-identification failure

Entity identification was not the dominant problem in the reviewed sample. Most sampled concepts were at least partially connected to source language.

### Entity-type/schema failure

This was the dominant problem.

The configuration forced domain concepts into four inappropriate generic categories. It accounts directly for 28 primary `WRONG_TYPE` classifications and also contributes to several duplicate and overgeneralized entities.

### Entity-resolution failure

Abbreviations, shortened names, singular/plural variants, and reordered personal names were not consistently merged.

This created multiple nodes for the same underlying concept.

### Relevance-filtering failure

Front matter, references, publication information, and broad contextual concepts contributed some entities that are correct but of limited value to the intended KG.

### Description-summarization failure

Some descriptions contain broader interpretations than the supporting source passage explicitly establishes.

---

## Severity assessment

**Entity identification:** Moderate quality
GraphRAG frequently recognized source-related concepts.

**Entity typing:** Poor quality / major bottleneck
The generic entity-type list is clearly unsuitable for this domain.

**Entity resolution:** Poor to moderate quality
Several important aliases and variants remain unmerged.

**Description reliability:** Moderate concern
Descriptions are useful summaries but cannot be treated as authoritative facts.

**Relevance filtering:** Moderate concern
Administrative, bibliographic, and overly broad concepts enter the graph.

---

## Methodological limitation

The reported percentages describe the **57-entity audit sample**, not a simple random estimate of all 312 entities.

The sample intentionally includes:

* Highly connected entities
* Stratified random entities from every raw type

Consequently, the percentages are appropriate for diagnosing pipeline weaknesses but should not be presented as exact population-wide error rates without reviewing a larger purely random sample.

The primary classifications are also mutually exclusive for counting purposes. Some entities have multiple simultaneous problems.

For example:

```text
PV-ANLAGEN
```

is both:

* Wrongly typed
* A duplicate/alias

It was assigned one primary classification while the additional problem was recorded in the audit notes and pipeline-stage fields.

---

## Stage 3 conclusion

GraphRAG generally identified concepts related to the Austrian Photovoltaic Strategy, with 48 of 57 sampled entities directly supported and nine partially supported.

However, successful identification did not translate into a controlled or reliable domain graph. Only 13 sampled entities were classified as both correct and relevant without a more important primary problem.

The dominant failure is entity typing. The generic categories `ORGANIZATION`, `PERSON`, `GEO`, and `EVENT` are inadequate for representing solar-market policies, technologies, infrastructure, targets, metrics, stakeholder groups, regulatory mechanisms, and environmental factors.

Entity resolution is a second major weakness. Important abbreviations and variants such as BMK, PV, strategy-name variants, and person-name variants remain separated.

The evidence supports treating GraphRAG as a useful but noisy entity-extraction layer. Experiment 2 should test domain-specific entity types and domain-relevant prompt examples. Nevertheless, a downstream controlled normalization and entity-resolution step will still be necessary even if raw extraction improves.
